packages/react-components/react-spinbutton/src/components/SpinButton/SpinButton.test.tsx

```diff
@@ -320,6 +348,97 @@ describe('SpinButton', () => {
+    fireEvent.keyDown(spinButton, { key: Keys.ArrowDown });
+    expect(onChange).not.toHaveBeenCalled();
+    fireEvent.keyDown(spinButton, { key: Keys.PageDown });
+    expect(onChange).not.toHaveBeenCalled();
+
+    userEvent.click(incrementButton);
+    // From 0 --> 1
+    expect(onChange).toHaveBeenCalledTimes(1);
+
+    userEvent.click(incrementButton);
+    // At max bound, no change
+    expect(onChange).toHaveBeenCalledTimes(1);
+    fireEvent.keyDown(spinButton, { key: Keys.ArrowUp });
+    expect(onChange).toHaveBeenCalledTimes(1);
+    fireEvent.keyDown(spinButton, { key: Keys.PageUp });
+    expect(onChange).toHaveBeenCalledTimes(1);
+  });
+
+  it('does not call change at bounds when controlled', () => {
+    const onChange = jest.fn();
+
+    const { getAllByRole, rerender } = render(<SpinButton value={0} min={0} max={1} onChange={onChange} />);
+
+    const [incrementButton, decrementButton] = getAllByRole('button');
+    const spinButton = getSpinButtonInput();
+
+    userEvent.click(decrementButton);
+    // Already at min bound, no change
+    expect(onChange).not.toHaveBeenCalled();
+    fireEvent.keyDown(spinButton, { key: Keys.ArrowDown });
+    expect(onChange).not.toHaveBeenCalled();
+    fireEvent.keyDown(spinButton, { key: Keys.PageDown });
+    expect(onChange).not.toHaveBeenCalled();
+
+    rerender(<SpinButton value={1} min={0} max={1} onChange={onChange} />);
+
+    userEvent.click(incrementButton);
+    // Already at maz bound, no change
+    expect(onChange).not.toHaveBeenCalled();
+    fireEvent.keyDown(spinButton, { key: Keys.ArrowUp });
+    expect(onChange).not.toHaveBeenCalled();
+    fireEvent.keyDown(spinButton, { key: Keys.PageUp });
+    expect(onChange).not.toHaveBeenCalled();
+  });
+
+  it('does not call on change when defaultValue is `null` when uncontrolled', () => {
+    const onChange = jest.fn();
+    const { getAllByRole } = render(<SpinButton defaultValue={null} onChange={onChange} />);
+
+    const [incrementButton, decrementButton] = getAllByRole('button');
+    userEvent.click(incrementButton);
+
+    expect(onChange).not.toHaveBeenCalled();
+    expect(getSpinButtonInput().value).toBe('');
+
+    userEvent.click(decrementButton);
+
+    expect(onChange).not.toHaveBeenCalled();
+    expect(getSpinButtonInput().value).toBe('');
+  });
+
+  it('does not call on change when value is `null` when controlled', () => {
+    const onChange = jest.fn();
+    const { getAllByRole, rerender } = render(<SpinButton value={null} onChange={onChange} />);
+
+    const [incrementButton, decrementButton] = getAllByRole('button');
+    userEvent.click(incrementButton);
+
+    expect(onChange).not.toHaveBeenCalled();
+    expect(getSpinButtonInput().value).toBe('');
+
+    rerender(<SpinButton value={null} onChange={onChange} />);
+    userEvent.click(decrementButton);
+
+    expect(onChange).not.toHaveBeenCalled();
+    expect(getSpinButtonInput().value).toBe('');
   });
```

Comment:
How does a user get out of a null value?

I'm considering the case where the SpinButton is empty and the user doesn't know what to type into the SpinButton manually so they try stepping instead... It looks like nothing will happen now?

When in this case, how would the creator know that the user is trying to set the SpinButton unless the explicitly type into the SpinButton?

I feel like it should fire on increment/decrement, maybe with the value of 1/-1 respectively? That would allow the user to to get out of the null case by stepping in the uncontrolled case, while also allowing the creator to know what the user's intentions are (at that point it's up to the creator to change it to something non-null, or potentially be notified again with a 1/-1 value).

For non-step cases (e.g. ENTER/blur) I would expect onChange to not fire if the value remained empty

Reply:
> I'm considering the case where the SpinButton is empty and the user doesn't know what to type into the SpinButton manually (or the SpinButton is non-editable, (e.g. does not allow freeform input)... is that even possible today, maybe the creator could pass readonly to the input???) so they try stepping instead... It looks like nothing will happen now?

An author could use slots to disable the input but we've decided against disallowing freeform input as part of SpinButton's API (it was there previously) because it makes the control inaccessible (there's no indication the input field disallows input). Slots give authors a lot of power when using components and I'm hesitant to make changes to the built-in behavior just because someone could do something. In this scenario, I think my response is, "don't disable the input as you're making the SpinButton inaccessible".

> When in this case, how would the creator know that the user is trying to set the SpinButton unless the explicitly type into the SpinButton?

Today you'd have to use the slots API to attach listeners to the input field, increment button and decrement button. I've opened a [feature request about an "intent to change" event handler](https://github.com/microsoft/fluentui/issues/23043)

> I feel like it should fire on increment/decrement, maybe with the value of 1 / -1 respectively? That would allow the user to to get out of the null case by stepping in the uncontrolled case, while also allowing the creator to know what the user's intentions are (at that point it's up to the creator to change it to something non-null, or potentially be notified again with a 1 / -1 value).Note, that is if no min is provided, if a min is provided you could return min+1 / min

Would this apply to the uncontrolled case only?

Since the only way for SpinButton to have its value set to `null` is for it to be set via props (SpinButton will never set null internally) I wonder if it's better to remove `| null` from the type of `defaultValue`?

Reply:
I spent some time thinking this over and I agree with your assessment. I've updated the null behavior to work with increment/decrement buttons.